### DISCUSSION TITLE
About hyperparameters

### DISCUSSION THREAD

**Discussion by rabiyaabbasi** [ultralytics/yolov5#7319] [Q&A]
Hi, I am new to YOLOV5, so I have a couple of questions.

There are five hyp.yaml files in hyps directory (hyp.Objects365.yaml, hyp.scratch-low.yaml, hyp.scratch-high.yaml, hyp.scratch-med.yaml,hyp.VOC.yaml). My first question is what is the difference between all of them particularly  hyp.scratch-low.yaml, hyp.scratch-high.yaml, and  hyp.scratch-med.yaml. I know some values of parameters are changed but broadly whats the difference? And they will impact my model?

I am using the pre-trained model to train my custom data using command (!python train.py --img 416 --batch 20 --epochs 300 --data dataset.yaml --weights yolov5s.pt --cache). My question is why the default hyperparameter file for training is  hyp.scratch-low.yaml? Can I change it to medium or high? If so how? Secondly, I have seen with hyperparameter evolution we can optimise the hyperparameters but if I change  values of few hyperprameters in hyp.scratch-low.yaml file manually, will it bring changes to my model during training? 

Thanks, your guidance is highly appreciated. 

![Capture](https://user-images.githubusercontent.com/54917838/162048781-b049721f-6859-49f1-940c-0052613b7032.PNG)

**Comment by glenn-jocher** [comment]
@rabiyaabbasi 👋 Hello! Thanks for asking about **hyperparameters** that define training and augmentation settings. Higher hyperparameters are used for larger models to delay overfitting.

YOLOv5 🚀 applies online imagespace and colorspace augmentations in the trainloader (but not the val_loader) to present a new and unique augmented [Mosaic](https://github.com/ultralytics/yolov5/blob/90b7895d652c3bd3d361b2d6e9aee900fd67f5f7/utils/datasets.py#L678-L732) (original image + 3 random images) each time an image is loaded for training. **Images are never presented twice in the same way**.

![YOLOv5 augmentation](https://user-images.githubusercontent.com/26833433/120995721-f3cfed00-c785-11eb-8ee2-b6ef2fa205e8.jpg)

### Augmentation Hyperparameters

The hyperparameters used to define these augmentations are in your hyperparameter file (default `data/hyp.scratch.yaml`) defined when training:
```
python train.py --hyp hyp.scratch-low.yaml
```
https://github.com/ultralytics/yolov5/blob/b94b59e199047aa8bf2cdd4401ae9f5f42b929e6/data/hyps/hyp.scratch-low.yaml#L6-L34

### Augmentation Previews

You can view the effect of your augmentation policy in your train_batch*.jpg images once training starts. These images will be in your train logging directory, typically `yolov5/runs/train/exp`:

`train_batch0.jpg` shows train batch 0 mosaics and labels:  
<img src="https://user-images.githubusercontent.com/26833433/131255960-b536647f-7c61-4f60-bbc5-cb2544d71b2a.jpg" width="800">


### YOLOv5 Albumentations Integration

[YOLOv5](https://github.com/ultralytics/yolov5) 🚀 is now fully integrated with [Albumentations](https://github.com/albumentations-team/albumentations), a popular open-source image augmentation package. Now you can train the world's best Vision AI models even better with custom Albumentations 😃! 

PR https://github.com/ultralytics/yolov5/pull/3882 implements this integration, which will automatically apply Albumentations transforms during YOLOv5 training if `albumentations>=1.0.3` is installed in your environment. See https://github.com/ultralytics/yolov5/pull/3882 for full details.

Example `train_batch0.jpg` on COCO128 dataset with Blur, MedianBlur and ToGray. See the YOLOv5 [Notebooks](https://github.com/ultralytics/yolov5/blob/master/tutorial.ipynb) to reproduce: <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>

<img src="https://user-images.githubusercontent.com/26833433/124400879-ff331b80-dd25-11eb-9b67-fe85ac4ca104.jpg" width="800">

Good luck 🍀 and let us know if you have any other questions!

**Comment by akashAD98** [comment]
@glenn-jocher its possible to apply augmention on specific class? i have 70 classes & i need to apply few augmentation on only specific to that class only, i dont want to apply all augmentions to all data

**Comment by 25benjaminli** [comment]
This is a great idea.

**Comment by Abhiramborige** [comment]
> 

thanks for the similar question. 
why the hyp.scratch-low.yaml has the values (cls and obj) high than the values present in hyp.scratch-high.yaml 
And also how should i set the values if I am in need of training on my custom dataset.
I wanted to detect small icons on a large images (80px size object on 2500px background)

PS: 
hyp.scratch-low.yaml contains: 
box: 0.05  # box loss gain
cls: 0.5  # cls loss gain
cls_pw: 1.0  # cls BCELoss positive_weight
obj: 1.0  # obj loss gain (scale with pixels)
iou_t: 0.20  # IoU training threshold

hyp.scratch-high.yaml contains: 
box: 0.05  # box loss gain
cls: 0.3  # cls loss gain
cls_pw: 1.0  # cls BCELoss positive_weight
obj: 0.7  # obj loss gain (scale with pixels)
obj_pw: 1.0  # obj BCELoss positive_weight
iou_t: 0.20  # IoU training threshold

Thankyou

**Comment by Seema1997** [comment]
Hi @glenn-jocher Thanks for asking about hyperparameters that define training and augmentation settings. Higher hyperparameters are used for larger models to delay overfitting.

YOLOv5 🚀 applies online imagespace and colorspace augmentations in the trainloader (but not the val_loader) to present a new and unique augmented [Mosaic](https://github.com/ultralytics/yolov5/blob/90b7895d652c3bd3d361b2d6e9aee900fd67f5f7/utils/datasets.py#L678-L732) (original image + 3 random images) each time an image is loaded for training. Images are never presented twice in the same way. 

This is the reply to one of the above-asked questions. My question is whether this answer is also valid for YOLOv7, YOLOv8, YOLOv9, and YOLOv10. Can you please give some highlights on the data augmentation technique used for  YOLOv7, YOLOv8, YOLOv9, and YOLOv10?
